src/gui/executionlogwidget.cpp

```diff
@@ -29,78 +29,73 @@
 #include "executionlogwidget.h"
 
 #include <QDateTime>
+#include <QMenu>
 #include <QPalette>
 
-#include "base/global.h"
-#include "loglistwidget.h"
+#include "log/logfiltermodel.h"
+#include "log/loglistview.h"
+#include "log/logmodel.h"
 #include "ui_executionlogwidget.h"
 #include "uithememanager.h"
 
-ExecutionLogWidget::ExecutionLogWidget(QWidget *parent, const Log::MsgTypes &types)
+ExecutionLogWidget::ExecutionLogWidget(const Log::MsgTypes types, QWidget *parent)
     : QWidget(parent)
     , m_ui(new Ui::ExecutionLogWidget)
-    , m_msgList(new LogListWidget(MAX_LOG_MESSAGES, types, this))
-    , m_peerList(new LogListWidget(MAX_LOG_MESSAGES, Log::ALL, this))
+    , m_messageFilterModel(new LogFilterModel(types, this))
 {
     m_ui->setupUi(this);
 
+    LogMessageModel *messageModel = new LogMessageModel(this);
+    m_messageFilterModel->setSourceModel(messageModel);
+    LogListView *messageView = new LogListView(this);
+    messageView->setModel(m_messageFilterModel);
+    messageView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(messageView, &LogListView::customContextMenuRequested, this, [this, messageView, messageModel](const QPoint &pos)
+    {
+        displayContextMenu(pos, messageView, messageModel);
+    });
+
+    LogPeerModel *peerModel = new LogPeerModel(this);
+    LogListView *peerView = new LogListView(this);
+    peerView->setModel(peerModel);
+    peerView->setContextMenuPolicy(Qt::CustomContextMenu);
+    connect(peerView, &LogListView::customContextMenuRequested, this, [this, peerView, peerModel](const QPoint &pos)
+    {
+        displayContextMenu(pos, peerView, peerModel);
+    });
+
+    m_ui->tabGeneral->layout()->addWidget(messageView);
+    m_ui->tabBan->layout()->addWidget(peerView);
+
 #ifndef Q_OS_MACOS
     m_ui->tabConsole->setTabIcon(0, UIThemeManager::instance()->getIcon("view-calendar-journal"));
     m_ui->tabConsole->setTabIcon(1, UIThemeManager::instance()->getIcon("view-filter"));
 #endif
-    m_ui->tabGeneral->layout()->addWidget(m_msgList);
-    m_ui->tabBan->layout()->addWidget(m_peerList);
-
-    const Logger *const logger = Logger::instance();
-    for (const Log::Msg &msg : asConst(logger->getMessages()))
-        addLogMessage(msg);
-    for (const Log::Peer &peer : asConst(logger->getPeers()))
-        addPeerMessage(peer);
-    connect(logger, &Logger::newLogMessage, this, &ExecutionLogWidget::addLogMessage);
-    connect(logger, &Logger::newLogPeer, this, &ExecutionLogWidget::addPeerMessage);
 }
 
 ExecutionLogWidget::~ExecutionLogWidget()
 {
     delete m_ui;
 }
 
-void ExecutionLogWidget::showMsgTypes(const Log::MsgTypes &types)
+void ExecutionLogWidget::setMessageTypes(const Log::MsgTypes types)
 {
-    m_msgList->showMsgTypes(types);
+    m_messageFilterModel->setMessageTypes(types);
 }
 
-void ExecutionLogWidget::addLogMessage(const Log::Msg &msg)
+void ExecutionLogWidget::displayContextMenu(const QPoint &pos, const LogListView *view, const BaseLogModel *model) const
```

Comment:
Seems this doesn't need to be in class, it should be moved to anonymous namespace.

Reply:
This internally calls tr, and I was once asked to only call tr of concerned class not QObject::tr directly, otherwise my first thought was to make it free function as well.

Reply:
QObject::tr in free functuon is bad since it doesn't produce any meaningful context.

Reply:
ok, forget it.